Zend/zend_compile.c

```diff
@@ -1403,26 +1413,31 @@ static zend_bool zend_try_ct_eval_class_const(zval *zv, zend_string *class_name,
 {
 	uint32_t fetch_type = zend_get_class_fetch_type(class_name);
 	zval *c;
+	zend_class_entry *ce = NULL;
+	zend_class_constant_info *const_info;
+
+	if (CG(compiler_options) & ZEND_COMPILE_NO_PERSISTENT_CONSTANT_SUBSTITUTION) {
+		return 0;
+	}
 
 	if (class_name_refers_to_active_ce(class_name, fetch_type)) {
-		c = zend_hash_find(&CG(active_class_entry)->constants_table, name);
+		ce = CG(active_class_entry);
 	} else if (fetch_type == ZEND_FETCH_CLASS_DEFAULT && !(CG(compiler_options) & ZEND_COMPILE_NO_CONSTANT_SUBSTITUTION)) {
-		zend_class_entry *ce = zend_hash_find_ptr_lc(CG(class_table), ZSTR_VAL(class_name), ZSTR_LEN(class_name));
-		if (ce) {
-			c = zend_hash_find(&ce->constants_table, name);
-		} else {
+		ce = zend_hash_find_ptr_lc(CG(class_table), ZSTR_VAL(class_name), ZSTR_LEN(class_name));
+	}
+
+	if (ce != NULL && (const_info = zend_hash_find_ptr(&ce->constants_info, name)) != NULL) {
+		c = OBJ_CONST_NUM(const_info->ce, const_info->offset);
+
+		if (c == NULL || Z_TYPE_P(c) > IS_OBJECT) {
 			return 0;
 		}
-	} else {
-		return 0;
-	}
 
-	if (CG(compiler_options) & ZEND_COMPILE_NO_PERSISTENT_CONSTANT_SUBSTITUTION) {
-		return 0;
-	}
+		if (!zend_verify_const_access(const_info, EG(current_execute_data) ? EG(scope) : CG(active_class_entry))) {
```

Comment:
Given this code is in zend_compile, should this always use `CG(active_class_entry)`?